stan/math/prim/fun/grad_2F1.hpp

```diff
@@ -9,144 +9,327 @@
+ * to within <code>precision</code> or throwing when the
+ * function takes <code>max_steps</code> steps.
+ *
+ * Overload for use where the destination gradients are not required to be the
+ * same type as the input variables (most use-cases except grad_inc_beta)
+ *
+ * @tparam ReturnSameT Whether return gradients need to be the same type as
+ *  as inputs
+ * @tparam T1 scalar type of a1
+ * @tparam T2 scalar type of a2
+ * @tparam T3 scalar type of b1
+ * @tparam T_z scalar type of z
+ * @param[in] a1 see generalized hypergeometric function definition
+ * @param[in] a2 see generalized hypergeometric function definition
+ * @param[in] b1 see generalized hypergeometric function definition
+ * @param[in] z see generalized hypergeometric function definition
+ * @param[in] precision magnitude of the increment of the infinite sum
+ *   to truncate the sum
+ * @param[in] max_steps number of steps to take
+ */
+template <bool ReturnSameT, typename T1, typename T2, typename T3, typename T_z,
+          require_not_t<std::integral_constant<bool, ReturnSameT>>* = nullptr>
+auto grad_2F1(const T1& a1, const T2& a2, const T3& b1, const T_z& z,
+              double precision = 1e-14, int max_steps = 1e6) {
+  return internal::grad_2F1_impl<
+      !is_constant<T1>::value, !is_constant<T2>::value, !is_constant<T3>::value,
+      !is_constant<T_z>::value>(value_of(a1), value_of(a2), value_of(b1),
+                                value_of(z), precision, max_steps);
+}
+
+/**
+ * Calculate the gradients of the hypergeometric function (2F1)
+ * as the power series stopping when the series converges
+ * to within <code>precision</code> or throwing when the
+ * function takes <code>max_steps</code> steps.
+ *
+ * Overload for use where the destination gradients should be the same type
+ * as the input variables (needed for the grad_inc_beta overloads)
+ *
+ * @tparam ReturnSameT Whether return gradients need to be the same type as
+ *  as inputs
+ * @tparam T1 scalar type of a1
+ * @tparam T2 scalar type of a2
+ * @tparam T3 scalar type of b1
+ * @tparam T_z scalar type of z
+ * @param[in] a1 see generalized hypergeometric function definition
+ * @param[in] a2 see generalized hypergeometric function definition
+ * @param[in] b1 see generalized hypergeometric function definition
+ * @param[in] z see generalized hypergeometric function definition
+ * @param[in] precision magnitude of the increment of the infinite sum
+ *   to truncate the sum
+ * @param[in] max_steps number of steps to take
+ */
+template <bool ReturnSameT, typename T1, typename T2, typename T3, typename T_z,
+          require_t<std::integral_constant<bool, ReturnSameT>>* = nullptr>
+auto grad_2F1(const T1& a1, const T2& a2, const T3& b1, const T_z& z,
+              double precision = 1e-14, int max_steps = 1e6) {
+  return internal::grad_2F1_impl<true, true, true, true>(a1, a2, b1, z,
+                                                         precision, max_steps);
+}
```

Comment:
If you set a default value for `ReturnSameT` like `bool ReturnSameT = false` do you need the other specialization below this one?

Reply:
Possibly, but since the default values have to come last in the template list I'd have to make some changes to how function would get called. I might leave that as a future TODO